app/controllers/api/api_controller.rb

```diff
@@ -0,0 +1,38 @@
+class API::APIController < ApplicationController
+  layout false
+
+  protected
+
+  def error(message, status = 500)
```

Comment:
> It is better to raise an exception and then have a handler on the controller that renders the error.

I think that's exactly what happens here. E.g. if you take a look at new import controller:
https://github.com/concord-consortium/lara/pull/349/files#diff-9a67abf55ca74f95d468cc674d3f0be8
There are some possible exceptions, controller handles them and renders an error. The only difference is that it uses the helper method `error` instead of crafting json by hand. Only to avoid the current situation where every single controller in api/v1 uses different json format for errors.

I could call this method `renderErrorJson` to be more explicit about render action. Or just generate JSON and call `render` in the child controller itself. But it all seems to be the same. If developer wants it, he will call render twice anyway. :wink: 

But I guess I'm not understating what you actually meant. Maybe you could use this tiny import controller as an example?

Reply:
So in that import controller it is calling the error method defined here.  Instead it could raise an exception.  It is also calling not_authorized, but instead could just let the exception go unhandled in the action and let a handler in the api_controller take care of this.

This is done using rescue_from: https://guides.rubyonrails.org/v3.2/action_controller_overview.html#rescue_from

So to be specific:
- the line which rescues the CanCan::AccessDenied exception can be removed from the import controller.  And a rescue_from line can be added to the API controller.  The exception is passed as an argument to the rescue_from handler.
- the `error(import_result[:error], 500)` can be replaced with something like `raise APIError(import_result[:error])` where the import error is a new type. defined by the import controller. And a rescue_from line is added to the API controller to handle these APIErrors.

An example of why this is useful. On this line:
https://github.com/concord-consortium/lara/pull/349/files#diff-3593224c23f37a55df736f8254328a52R6
If the record is not found Rails will raise an exception, with rescue_from this could be caught and a nice JSON response can be returned (instead of the default html response).  Instead of using rescue_from a developer might fix this using an error call. And they might assume they can just call error and not worry about it. So they might replace this line with:
```
begin
  Sequence.find(params[:id]).destroy
rescue RecordNotFound
   error(...)
end
```
Because there is a following `render` call at the end of the action, if this RecordNotFound is raised then there will be a double render error. 

We had this problem in the Portal, and it was pretty hard to find.  My solution there was to just replace all calls to `error(...)` with `return error(...)`.  But raising exceptions and handing them with rescue_from seems to be a better way to do this. It prevents the double render, and also means we shouldn't need to put as much error handling in each action.

Reply:
Thanks! I was just missing `rescue_from`. I've reimplemented new controllers to use it, and added some test to verify that it works as expected (some old tests were already checking 403 case).